presto-hive/src/main/java/io/prestosql/plugin/hive/ViewReaderUtil.java

```diff
@@ -0,0 +1,229 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static org.apache.hadoop.hive.metastore.TableType.VIRTUAL_VIEW;
+
+public final class ViewReaderUtil
+{
+    private ViewReaderUtil()
+    {}
+
+    public interface ViewReader
+    {
+        ConnectorViewDefinition decodeViewData(String viewData, Table table, CatalogName catalogName);
+    }
+
+    public static ViewReader createViewReader(SemiTransactionalHiveMetastore metastore, HiveIdentity identity, Table table, TypeManager typemanager)
+    {
+        if (isPrestoView(table)) {
+            return new io.prestosql.plugin.hive.ViewReaderUtil.PrestoViewDecoder();
+        }
+        else {
+            return new io.prestosql.plugin.hive.ViewReaderUtil.HiveViewDecoder(new CoralSemiTransactionalHiveMSCAdapter(metastore, identity), typemanager);
+        }
+    }
+
+    public static final String PRESTO_VIEW_FLAG = "presto_view";
+    static final String VIEW_PREFIX = "/* Presto View: ";
+    static final String VIEW_SUFFIX = " */";
+    private static final JsonCodec<ConnectorViewDefinition> VIEW_CODEC =
+            new JsonCodecFactory(new ObjectMapperProvider()).jsonCodec(ConnectorViewDefinition.class);
+    private static Logger log = Logger.get(io.prestosql.plugin.hive.ViewReaderUtil.class);
+
+    public static boolean isPrestoView(Table table)
+    {
+        return "true".equals(table.getParameters().get(PRESTO_VIEW_FLAG));
+    }
+
+    public static boolean isHiveOrPrestoView(Table table)
+    {
+        return table.getTableType().equals(TableType.VIRTUAL_VIEW.name());
+    }
+
+    public static boolean canDecodeView(Table table)
+    {
+        // we can decode Hive or Presto view
+        return table.getTableType().equals(VIRTUAL_VIEW.name());
+    }
+
+    public static String encodeViewData(ConnectorViewDefinition definition)
+    {
+        byte[] bytes = VIEW_CODEC.toJsonBytes(definition);
+        String data = Base64.getEncoder().encodeToString(bytes);
+        return VIEW_PREFIX + data + VIEW_SUFFIX;
+    }
+
+    /**
+     * Supports decoding of Presto views
+     */
+    public static class PrestoViewDecoder
+            implements ViewReader
+    {
+        @Override
+        public ConnectorViewDefinition decodeViewData(String viewData, Table table, CatalogName catalogName)
+        {
+            checkCondition(viewData.startsWith(VIEW_PREFIX), HIVE_INVALID_VIEW_DATA, "View data missing prefix: %s", viewData);
+            checkCondition(viewData.endsWith(VIEW_SUFFIX), HIVE_INVALID_VIEW_DATA, "View data missing suffix: %s", viewData);
+            viewData = viewData.substring(VIEW_PREFIX.length());
+            viewData = viewData.substring(0, viewData.length() - VIEW_SUFFIX.length());
+            byte[] bytes = Base64.getDecoder().decode(viewData);
+            return VIEW_CODEC.fromJson(bytes);
+        }
+    }
+
+    /**
+     * Class to decode Hive view definitions
+     */
+    public static class HiveViewDecoder
```

Comment:
Do we expect this class to get substantially bigger as the transactional support is expanded? If so, it would be better to make it a top level class.